### DISCUSSION TITLE
Simulated environmental testing

### DISCUSSION THREAD

**Discussion by wChenonly** [vitest-dev/vitest#6197] [Q&A]
Some of the data I need for testing, methods are sdk methods provided by our own app container, the project is running in the container's webview, the
For example

// init.ts
```
import { initDeviceReady } from '@xxx/sdk'
export async function wup_initReady() {
  const res = initDeviceReady()
  return res
}
```

I'm writing unit tests now, and some of the methods don't run on the node side or the pure web side. I'm using mock
// init.test.ts
```
import {wup_initReady} from './init'
vi.mock('@xx/sdk', () => ({
  initDeviceReady: vi.fn(() => ({ retCode: '000000', retData: null, retInfo: 'success' }))
}))
expect(await wup_initReady()).toBe(true)
```

But it's a pain in the ass. I need to run our app container first to get the res of initDeviceReady, then put it in vi.fn, and run vitest after that.

I'm wondering if there's a way to run our app container first, store the initDeviceReady return value, and then when I run vitest, I can just use that value without having to mock it again.

But our app container is running on the phone, not on the node, so I don't know if there's a way to store the value and then use it on the node.

Or can custom-environment https://vitest.dev/guide/environment.html#custom-environment directly emulate your own app container?

Or any other simple programs would be appreciated!

**Comment by sheremet-va** [comment]
> I'm wondering if there's a way to run our app container first, store the initDeviceReady return value, and then when I run vitest, I can just use that value without having to mock it again.

You can run the container in [`globalSetup`](https://vitest.dev/config/#globalsetup) and pass down the values to your tests:

```ts
// globalSetup.js
export default function setup({ provide }) {
  const data = await initData()
  provide('sdk_data', data)
}
```

```ts
// setupFile.js
import { inject } from 'vitest'
vi.mock('../sdk', () => {
  return inject('sdk_data')
})
```

**Reply by wChenonly** [reply]
Thank you for your reply, but I still don't quite understand, the container I'm referring to is a hybrid development application, I embedded h5 pages, the h5 pages rely on some methods provided by application->sdk, such as one of the @xxx/api package initDeviceReady, and the container use I'm running is vite dev

**Reply by wChenonly** [reply]
@sheremet-va Can you help me in looking at this. I still don't understand it very well.
